# DMA Assignment 1 - Data Preprocessing

Data transformations are useful for preparing a dataset for answering a particular question. Part of this process involves generating features from the dataset you find relevant to the question at hand. For this lab, we will be using a Yelp reviews dataset. Each row in the dataset depicts one review along with the features of the review (the reviewer, the review text, etc.). The goal of this lab is to convert this reviews dataset into a reviewers dataset by creating different features of each reviewer.

The submission for this assignment should be done *individually*, although you are allowed to work in groups of 2.



# Environment Setup
Run this cell to setup your environment.

In [0]:
# Importing libraries
import pandas as pd
import math
import numpy as np
print('Libraries Imported')

#DOWNLOADING DATASET
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
!unzip yelp_reviews.zip
print('Dataset Downloaded: yelp_reviews.csv')
df=pd.read_csv('yelp_reviews.csv')
print(df.head())

print('Setup Complete')

Libraries Imported
--2019-02-06 05:56:00--  http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 376638166 (359M) [text/csv]
Saving to: ‘yelp_reviews.csv.1’

yelp_reviews.csv.1  100%[===================>] 359.19M  5.91MB/s    in 2m 9s   

2019-02-06 05:58:09 (2.79 MB/s) - ‘yelp_reviews.csv.1’ saved [376638166/376638166]

unzip:  cannot find or open yelp_reviews.zip, yelp_reviews.zip.zip or yelp_reviews.zip.ZIP.
Dataset Downloaded: yelp_reviews.csv
     type             business_id                 user_id  stars  \
0  review  mxrXVZWc6PWk81gvOVNOUw  mv7shusL4Xb6TylVYBv4CA      4   
1  review  mxrXVZWc6PWk81gvOVNOUw  0aN5QPhs-VwK2vusKG0waQ      5   
2  review  kK4AzZ0YWI-U2G-paAL7Fg  0aN5QPhs-VwK2vusKG0waQ      5   
3  review  mxrX

**Q1: What was the highest number of reviews for any one business_id?**
- For this task, we will need to group the reviews dataset by business_id. This will aggregate data for each business, which is what we need. This can be done using [groupby](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html).
    - ```yelp_businesses = yelp_dataset.groupby('business_id').size() ```
    - The .size() counts the number of instances for each business_id, which gives us the number of reviews as each instance in this dataset is a review.
    - The following command will sort this list, after which you can take note of the highest value: ```sorted_yelp_businesses = yelp_businesses.sort(ascending=False,inplace=False)```
    - This approach allows you to see the data structure being used in the sort. A quicker approach to getting the max would be to use the max function: ```max(yelp_businesses)```
<br>
<br>

In [0]:
# YOUR CODE HERE
df.groupby('business_id').size().sort_values(ascending=False, inplace=False).max()

4128

###Answer:4128

**Q2: What was the average number of reviews for a business_id?**

In [0]:
# YOUR CODE HERE

df.groupby("business_id").size().mean()

12.63413902163123

###Answer:12.63413902163123

**Q3: What is the average number of reviews per reviewer?**

In [0]:
# YOUR CODE HERE

df.groupby("user_id").size().mean()

3.188511934933203

###Answer:3.188511934933203

**Q4: What is the average number of cool votes per reviewer?**

In [0]:
# YOUR CODE HERE

df[["cool_votes","user_id"]].groupby("user_id").sum().mean()

cool_votes    1.241728
dtype: float64

###Answer: 1.241728

**Q5: What is the average number of funny votes per reviewer?**

In [0]:
# YOUR CODE HERE


df[["funny_votes","user_id"]].groupby("user_id").sum().mean()

funny_votes    1.101265
dtype: float64

###Answer:1.101265

**Q6: What is the average number of useful votes per reviewer?**

In [0]:
# YOUR CODE HERE

df[["useful_votes","user_id"]].groupby("user_id").sum().mean()

useful_votes    2.484476
dtype: float64

###Answer:2.484476

**Q7: What is the average of the log of the number of reviews per reviewer?**

In [0]:
# YOUR CODE HERE

np.log(df.groupby("user_id").size()).mean()

0.7681766036714884

###Answer:0.7681766036714884

**Q8: What is the average of the log of the number of cool votes per reviewer?**

In [0]:
# YOUR CODE HERE

np.log(df[["cool_votes","user_id"]].groupby("user_id").sum()).mean()

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in log
  


cool_votes   -inf
dtype: float64

###Answer:-inf

**Q9: What is the average of the log of the number of funny votes per reviewer?**

In [0]:
# YOUR CODE HERE


np.log(df[["funny_votes","user_id"]].groupby("user_id").sum()).mean()

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:3: RuntimeWarning: divide by zero encountered in log
  This is separate from the ipykernel package so we can avoid doing imports until


funny_votes   -inf
dtype: float64

###Answer: -inf

**Q10: What is the average of the log of the number of useful votes per reviewer?**

In [0]:
# YOUR CODE HERE



np.log(df[["useful_votes","user_id"]].groupby("user_id").sum()).mean()

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: divide by zero encountered in log
  after removing the cwd from sys.path.


useful_votes   -inf
dtype: float64

###Answer:-inf

**Q11: Find the average of the percentage of total cool votes out of total votes for each reviewer.**

In [0]:
# YOUR CODE HERE
df["total"]=df["cool_votes"]+df["useful_votes"]+df["funny_votes"]
df[["total","user_id"]].groupby("user_id").sum()
a=df[["cool_votes","user_id"]].groupby("user_id").sum()
b=df[["total","user_id"]].groupby("user_id").sum()
np.mean(a["cool_votes"]/b["total"])

0.1927040524186044

###Answer:19.27040524186044%

**Q12: Find the average of the percentage of total funny votes out of total votes for each reviewer.**

In [0]:
# YOUR CODE HERE

df["total"]=df["cool_votes"]+df["useful_votes"]+df["funny_votes"]
df[["total","user_id"]].groupby("user_id").sum()
a=df[["funny_votes","user_id"]].groupby("user_id").sum()
b=df[["total","user_id"]].groupby("user_id").sum()
np.mean(a["funny_votes"]/b["total"])


0.18256423845801606

###Answer:18.256423845801606%

**Q13: Find the average of the percentage of total useful votes out of total votes for each reviewer.**

In [0]:
# YOUR CODE HERE

df["total"]=df["cool_votes"]+df["useful_votes"]+df["funny_votes"]
df[["total","user_id"]].groupby("user_id").sum()
a=df[["useful_votes","user_id"]].groupby("user_id").sum()
b=df[["total","user_id"]].groupby("user_id").sum()
np.mean(a["useful_votes"]/b["total"])*100


62.47317091233795

###Answer:62.47317091233795%

**Q14: Average review text length (in non-space characters)**

In [0]:
# YOUR CODE HERE
q=0
for i in df["text"]:
  q+=len(i) - i.count(' ')
q*1.0/len(df["text"])

499.18610090393645

###Answer:499.18610090393645

**Q15: Year in which the reviewer wrote the most reviews. Once you have this for each reviewer, subtract the minimum possible year (2005) from each so that your final feature values are 0, 1, 2 etc.**

In [0]:
# YOUR CODE HERE
p=[]
import re
for i in df["date"]:
  p+=re.findall("[\d]{4}",i)
df["year"]=p
max_year=df.groupby(["user_id","year"]).size().to_frame().reset_index().groupby(["user_id",0]).max()
[int(i)-2005 for i in max_year.values]


[6,
 9,
 3,
 11,
 10,
 11,
 9,
 11,
 10,
 9,
 9,
 9,
 10,
 10,
 2,
 6,
 8,
 9,
 8,
 7,
 7,
 6,
 5,
 4,
 10,
 10,
 5,
 8,
 10,
 10,
 8,
 10,
 11,
 6,
 8,
 9,
 8,
 7,
 11,
 11,
 10,
 10,
 9,
 7,
 9,
 8,
 10,
 10,
 11,
 9,
 9,
 8,
 9,
 10,
 5,
 6,
 10,
 9,
 11,
 10,
 10,
 11,
 9,
 10,
 10,
 9,
 10,
 11,
 9,
 10,
 10,
 9,
 10,
 8,
 7,
 4,
 10,
 11,
 6,
 9,
 10,
 10,
 10,
 10,
 9,
 11,
 10,
 4,
 11,
 9,
 9,
 10,
 10,
 6,
 9,
 7,
 8,
 6,
 4,
 11,
 7,
 10,
 10,
 7,
 5,
 11,
 9,
 11,
 10,
 11,
 10,
 10,
 10,
 9,
 5,
 11,
 7,
 5,
 8,
 8,
 5,
 10,
 0,
 11,
 10,
 7,
 9,
 9,
 11,
 10,
 9,
 7,
 5,
 10,
 5,
 10,
 11,
 11,
 10,
 9,
 8,
 7,
 9,
 8,
 9,
 8,
 5,
 11,
 10,
 11,
 11,
 10,
 10,
 8,
 9,
 6,
 9,
 10,
 10,
 7,
 8,
 4,
 10,
 11,
 3,
 11,
 11,
 7,
 5,
 9,
 11,
 7,
 3,
 8,
 6,
 7,
 10,
 9,
 11,
 7,
 8,
 11,
 9,
 10,
 11,
 11,
 7,
 8,
 7,
 10,
 8,
 5,
 5,
 6,
 10,
 11,
 7,
 10,
 10,
 5,
 6,
 7,
 6,
 7,
 9,
 6,
 10,
 9,
 11,
 8,
 9,
 11,
 10,
 11,
 10,
 10,
 5,
 4,
 10,
 6,
 10,
 7,
 10,
 7,
 9,
 

###Answer: 
[6,
 9,
 3,
 11,
 10,
 11,
 9,
 11,
 10,
 9,
 9,
 9,
 10,
 10,
 2,
 6,
 8,
 9,
 8,
 7,
 7,
 6,
 5,
 4,
 10,
 10,
 5,
 8,
 10,
 10,
 8,
 10,
 11,
 6,
 8,
 9,
 8,
 7,
 11,
 11,
 10,
 10,
 9,
 7,
 9,
 8,
 10,
 10,
 11,
 9,
 9,
 8,
 9,
 10,
 5,
 6,
 10,
 9,
 11,
 10,
 10,
 11,
 9,
 10,
 10,
 9,
 10,
 11,
 9,
 10,
 10,
 9,
 10,
 8,
 7,
 4,
 10,
 11,
 6,
 9,
 10,
 10,
 10,
 10,
 9,
 11,
 10,
 4,
 11,
 9,
 9,
 10,
 10,
 6,
 9,
 7,
 8,
 6,
 4,
 11,
 7,
 10,
 10,
 7,
 5,
 11,
 9,
 11,
 10,
 11,
 10,
 10,
 10,
 9,
 5,
 11,
 7,
 5,
 8,
 8,
 5,
 10,
 0,
 11,
 10,
 7,
 9,
 9,
 11,
 10,
 9,
 7,
 5,
 10,
 5,
 10,
 11,
 11,
 10,
 9,
 8,
 7,
 9,
 8,
 9,
 8,
 5,
 11,
 10,
 11,
 11,
 10,
 10,
 8,
 9,
 6,
 9,
 10,
 10,
 7,
 8,
 4,
 10,
 11,
 3,
 11,
 11,
 7,
 5,
 9,
 11,
 7,
 3,
 8,
 6,
 7,
 10,
 9,
 11,
 7,
 8,
 11,
 9,
 10,
 11,
 11,
 7,
 8,
 7,
 10,
 8,
 5,
 5,
 6,
 10,
 11,
 7,
 10,
 10,
 5,
 6,
 7,
 6,
 7,
 9,
 6,
 10,
 9,
 11,
 8,
 9,
 11,
 10,
 11,
 10,
 10,
 5,
 4,
 10,
 6,
 10,
 7,
 10,
 7,
 9,
 11,
 6,
 9,
 8,
 8,
 11,
 11,
 10,
 9,
 9,
 10,
 9,
 6,
 11,
 10,
 8,
 9,
 8,
 9,
 8,
 10,
 11,
 11,
 11,
 8,
 6,
 9,
 9,
 11,
 11,
 8,
 11,
 10,
 11,
 11,
 10,
 8,
 10,
 6,
 11,
 10,
 9,
 9,
 8,
 10,
 8,
 7,
 11,
 8,
 9,
 8,
 7,
 6,
 6,
 9,
 5,
 7,
 10,
 8,
 9,
 8,
 11,
 10,
 10,
 7,
 9,
 5,
 11,
 9,
 8,
 7,
 9,
 5,
 11,
 10,
 10,
 5,
 11,
 9,
 10,
 9,
 11,
 10,
 9,
 11,
 9,
 10,
 5,
 6,
 4,
 10,
 7,
 3,
 11,
 7,
 10,
 11,
 10,
 8,
 8,
 8,
 3,
 10,
 9,
 9,
 8,
 10,
 9,
 7,
 8,
 10,
 7,
 11,
 10,
 7,
 10,
 10,
 11,
 7,
 10,
 7,
 11,
 4,
 5,
 10,
 9,
 2,
 10,
 7,
 9,
 11,
 10,
 10,
 10,
 4,
 11,
 8,
 10,
 10,
 8,
 11,
 11,
 8,
 10,
 7,
 10,
 11,
 11,
 10,
 9,
 10,
 9,
 10,
 8,
 10,
 8,
 6,
 6,
 8,
 9,
 10,
 6,
 11,
 6,
 10,
 10,
 6,
 7,
 8,
 8,
 10,
 11,
 10,
 11,
 10,
 11,
 8,
 10,
 8,
 6,
 10,
 9,
 11,
 6,
 6,
 11,
 10,
 9,
 8,
 9,
 11,
 6,
 11,
 10,
 10,
 11,
 11,
 8,
 11,
 9,
 10,
 5,
 6,
 11,
 10,
 11,
 10,
 11,
 7,
 7,
 10,
 11,
 9,
 7,
 9,
 11,
 3,
 4,
 7,
 11,
 6,
 9,
 11,
 6,
 6,
 8,
 9,
 11,
 10,
 7,
 10,
 9,
 10,
 10,
 7,
 10,
 11,
 11,
 9,
 2,
 11,
 8,
 10,
 11,
 6,
 9,
 10,
 8,
 10,
 8,
 11,
 7,
 11,
 9,
 9,
 7,
 10,
 9,
 11,
 10,
 8,
 6,
 8,
 9,
 11,
 10,
 11,
 7,
 11,
 10,
 5,
 7,
 11,
 11,
 8,
 10,
 9,
 11,
 4,
 10,
 9,
 11,
 10,
 10,
 10,
 8,
 9,
 8,
 11,
 4,
 11,
 8,
 10,
 5,
 9,
 11,
 10,
 3,
 8,
 10,
 2,
 11,
 9,
 10,
 9,
 5,
 9,
 10,
 11,
 11,
 9,
 11,
 10,
 11,
 10,
 8,
 11,
 10,
 11,
 8,
 9,
 11,
 6,
 10,
 7,
 10,
 9,
 10,
 5,
 11,
 6,
 10,
 3,
 4,
 9,
 6,
 10,
 5,
 10,
 7,
 10,
 8,
 3,
 2,
 11,
 11,
 9,
 8,
 9,
 9,
 9,
 6,
 10,
 11,
 10,
 7,
 8,
 9,
 11,
 9,
 8,
 11,
 7,
 10,
 9,
 10,
 11,
 9,
 8,
 9,
 10,
 11,
 10,
 10,
 11,
 9,
 4,
 10,
 5,
 8,
 10,
 6,
 5,
 10,
 9,
 9,
 10,
 11,
 10,
 1,
 6,
 3,
 8,
 11,
 4,
 3,
 11,
 10,
 9,
 10,
 9,
 4,
 5,
 11,
 10,
 7,
 7,
 10,
 11,
 9,
 6,
 9,
 7,
 11,
 10,
 10,
 5,
 3,
 10,
 9,
 5,
 9,
 10,
 9,
 10,
 10,
 8,
 10,
 10,
 6,
 10,
 4,
 6,
 3,
 11,
 11,
 2,
 9,
 6,
 9,
 8,
 9,
 9,
 5,
 9,
 11,
 11,
 9,
 9,
 10,
 9,
 11,
 11,
 9,
 11,
 7,
 8,
 9,
 5,
 4,
 9,
 11,
 8,
 5,
 7,
 4,
 4,
 5,
 9,
 7,
 6,
 5,
 2,
 9,
 11,
 10,
 9,
 8,
 7,
 10,
 9,
 11,
 10,
 7,
 8,
 10,
 7,
 10,
 9,
 10,
 11,
 8,
 7,
 9,
 8,
 11,
 11,
 9,
 10,
 10,
 10,
 11,
 10,
 5,
 11,
 9,
 8,
 6,
 3,
 10,
 9,
 11,
 11,
 9,
 9,
 11,
 7,
 10,
 11,
 11,
 10,
 9,
 8,
 8,
 9,
 11,
 10,
 11,
 11,
 7,
 6,
 11,
 9,
 10,
 10,
 9,
 8,
 10,
 9,
 8,
 10,
 6,
 9,
 11,
 8,
 9,
 5,
 9,
 11,
 10,
 9,
 4,
 10,
 8,
 11,
 10,
 9,
 8,
 10,
 6,
 11,
 10,
 4,
 4,
 9,
 10,
 8,
 10,
 9,
 8,
 9,
 10,
 11,
 10,
 7,
 8,
 2,
 7,
 11,
 9,
 8,
 11,
 10,
 3,
 11,
 10,
 6,
 9,
 9,
 8,
 6,
 6,
 7,
 10,
 6,
 9,
 8,
 9,
 8,
 11,
 6,
 7,
 10,
 3,
 10,
 10,
 11,
 10,
 11,
 9,
 10,
 9,
 9,
 11,
 10,
 8,
 7,
 6,
 11,
 7,
 2,
 10,
 9,
 10,
 9,
 10,
 9,
 10,
 8,
 11,
 10,
 11,
 10,
 2,
 11,
 7,
 9,
 7,
 9,
 8,
 10,
 6,
 9,
 10,
 9,
 7,
 10,
 10,
 5,
 4,
 10,
 7,
 7,
 10,
 9,
 7,
 11,
 11,
 8,
 11,
 9,
 10,
 9,
 9,
 10,
 9,
 8,
 7,
 10,
 6,
 8,
 7,
 11,
 6,
 8,
 10,
 9,
 8,
 10,
 6,
 10,
 9,
 10,
 9,
 5,
 8,
 10,
 7,
 10,
 9,
 10,
 10,
 6,
 9,
 10,
 8,
 10,
 6,
 10,
 10,
 7,
 3,
 11,
 10,
 9,
 11,
 4,
 6,
 7,
 10,
 10,
 8,
 9,
 11,
 11,
 9,
 11,
 8,
 7,
 6,
 5,
 11,
 4,
 6,
 11,
 5,
 10,
 8,
 11,
 10,
 5,
 9,
 11,
 10,
 4,
 9,
 7,
 8,
 8,
 8,
 9,
 10,
 11,
 10,
 9,
 10,
 8,
 8,
 10,
 10,
 7,
 11,
 10,
 10,
 11,
 7,
 6,
 3,
 8,
 11,
 10,
 10,
 6,
 11,
 9,
 5,
 8,
 7,
 11,
 7,
 8,
 8,
 7,
 9,
 10,
 ...]

**Q16: Come up with a new feature. This may be derived from existing features. Give your feature the name ** *my_new_feature.* **Dsiplay head() of this new feature.**

In [0]:
# YOUR CODE HERE

def percentage(x):
  return sum([1 for i in x if (i > 3)])*1.0/len(x)
percentage_rating=df.groupby("business_id")["stars"].apply(list).to_frame()
my_new_feature =percentage_rating["stars"].apply(percentage)
my_new_feature.head()

business_id
--5jkZ3-nUPZxUvtcbr8Uw    0.920000
--AKjxBmhm9DWrh-e0hTOw    1.000000
--BlvDO_RG2yElKu9XA1_g    0.900000
--Ol5mVSMaW8ExtmWRUmKA    1.000000
--Y_2lDOtVDioX5bwF6GIw    0.166667
Name: stars, dtype: float64

###Answer: My new feature is the percentage of good ratings for each business_id, considering 4 and 5 stars as good rating. 